File: binomial.py
from pysat.solvers import Glucose3

def generate_variables(n):
    variables = [[] for _ in range(n)]
    for i in range(n):
        for j in range(n):
            variables[i].append(i * n + j + 1)
    return variables

def at_most_one(clauses, variables):
    for i in range(0, len(variables)):
        for j in range(i + 1, len(variables)):
            clauses.append([-variables[i], -variables[j]])
    return clauses       

def exactly_one(clauses, variables):
    clauses.append(variables)
    at_most_one(clauses, variables)

def generate_clauses(n, variables):
    clauses = []
    
    #Exacly one queen per rows
    for i in range(n):
        exactly_one(clauses, variables[i])
    
    #Exacly one queen per columns
    for j in range(n):
        exactly_one(clauses, [variables[i][j] for i in range(n)])
        
    #At most one queen per diagonal
    for i in range(n):
        for j in range(n):
            for k in range(1, n):
                if i + k < n and j + k < n:
                    at_most_one(clauses, [variables[i][j], variables[i + k][j + k]])
                if i + k < n and j - k >= 0:
                    at_most_one(clauses, [variables[i][j], variables[i + k][j - k]])
    return clauses
         
def solve_n_queens(n):
    variables = generate_variables(n)
    clauses = generate_clauses(n, variables)
    print(clauses)

    solver = Glucose3()
    for clause in clauses:
        solver.add_clause(clause)

    if solver.solve():
        model = solver.get_model()
        return [[int(model[i * n + j] > 0) for j in range(n)] for i in range(n)]
    else:
        return None
            
def print_solution(solution):
    if solution is None:
        print("No solution found.")
    else:
        print(solution)
        for row in solution:
            print(" ".join("Q" if cell else "." for cell in row))


n = 8
# print(generate_clauses(n, generate_variables(n)))
solution = solve_n_queens(n)
print_solution(solution)

File: binomial.cpp
#include <iostream>
#include <vector>
#include <cstring>

std::vector<std::vector<int>> generate_variables(const int& n) {
    std::vector<std::vector<int>> variables;

    for (int i = 0; i < n; i++) {
        std::vector<int> row;
        for (int j = 0; j < n; j++) {
            row.push_back(i * n + j + 1); //Value in position (i, j) is (i * n + j + 1)
        }
        variables.push_back(row);
    }
    return variables;
}

void atMostOne(std::vector<std::vector<int>>& clauses, const std::vector<int>& variables) {
    for (int i = 0; i < variables.size(); i++) {
        for (int j = i + 1; j < variables.size(); j++) {
            std::vector<int> temp = {-variables[i], -variables[j]};
            clauses.push_back(temp);
        }
    }
}

void exactlyOne(std::vector<std::vector<int>>& clauses, const std::vector<int>& variables) {
    //At least a true value in variables
    clauses.push_back(variables);

    //Only one true value in variables
    atMostOne(clauses, variables);
}

std::vector<std::vector<int>> generate_clauses(const int& n, 
                                               const std::vector<std::vector<int>>& variables) {
    std::vector<std::vector<int>> clauses = {};

    //Exactly one queen in a row
    for (int i = 0; i < n; i++) {
        exactlyOne(clauses, variables[i]);
    }

    //Exactly one queen in a column
    for (int j = 0; j < n; j++) {
        std::vector<int> col;
        for (int i = 0; i < n; i++) {
            col.push_back(variables[i][j]);
        }
        exactlyOne(clauses, col);
    }

    //Only one queen in each diagonal
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            for (int k = 1; k < n; k++) {
                if (i + k < n && j + k < n) {
                    atMostOne(clauses, {variables[i][j], variables[i + k][j + k]});
                }
                if (i + k < n && j - k >= 0) {
                    atMostOne(clauses, {variables[i][j], variables[i + k][j - k]});
                }
            }
        }
    }
    return clauses;
}

void Solve() {
    int n = 5;
    std::vector<std::vector<int>> variables = generate_variables(n);
    std::vector<std::vector<int>> clauses = generate_clauses(n, variables);

    for (const auto& clause : clauses) {
        std::cout << "[";
        for (const auto& i : clause) {
            std::cout << i << " ";
        }
        std::cout << "0]\n";
    }
}

int main() {
    Solve();
    return 0;
}